### PR TITLE
br: support PiTR feature 

### PR DESCRIPTION
**Pull request by joccau**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/29501

Problem Summary:

### What is changed and how it works?
The new feature: PiTR
- Take a backup of key-value recored in TiKV If you  run `br log start ...`
- Restore table and record from snapshot-ts to restore-ts if you run `br restore point ...`

### Relative design/tests
- [Specification](https://docs.google.com/document/d/1nr4Ubd5oKBK1mY6eqJtktTtaFvZ_6HxUZhMUQIuCCcw/edit#heading=h.m5g2iqfkn37v)
- [Design](https://pingcap.feishu.cn/docs/doccnMvA0vKi8cnzIb2fAfUC6nd)
- [Test plan](https://docs.google.com/document/d/1tFXxjMQzLnSKf-MQyga7eb1H8DRmivD2kl3PQxpCVzM/edit?n=FD_342_Feature_test_plan)
  - [Issues record in development](https://docs.google.com/document/d/1pa-L1JV1W0fZRy44fAnSOEPjdSlxEVYZviDh7arTey8/edit#heading=h.ujsh83watf91)
  - [Base test cases](https://docs.google.com/spreadsheets/d/1ryPfNryaFl9o2_We0oGzmGa3os2aus-cv-tqgBpses0/edit#gid=0)
 

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [x] Performance regression: Consumes more CPU
- [x] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [x] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Support the new feature of PiTR.
```

### CODE REVIEW

File: br/pkg/utils/schema.go [hunk 1]
```diff
@@ -114,3 +114,8 @@ func GetSysDBName(tempDB model.CIStr) (string, bool) {
 	}
 	return tempDB.O[len(temporaryDBNamePrefix):], true
 }
+
+// EscapeName replaces all "`" in name with "``"
+func EscapeName(name string) string {
```
Comment (3pointer) [inline br/pkg/utils/schema.go:119]: This function can be removed.

File: br/cmd/br/stream.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
+// Copyright 2015 PingCAP, Inc.
```
Comment (wjhuang2016) [inline br/cmd/br/stream.go:1]: 2015?

File: br/pkg/gluetidb/glue.go [hunk 3]
```diff
@@ -106,12 +116,22 @@ func (g Glue) GetVersion() string {
...
 }
 
 func (gs *tidbSession) ExecuteInternal(ctx context.Context, sql string, args ...interface{}) error {
-	_, err := gs.se.ExecuteInternal(ctx, sql, args...)
+	rs, err := gs.se.ExecuteInternal(ctx, sql, args...)
+	// Some of SQLs (like ADMIN RECOVER INDEX) may lazily take effect
+	// when we polling the result set.
+	// At least call `next` once for triggering theirs side effect.
+	// (Maybe we'd better drain all returned rows?)
+	if rs != nil {
```
Comment (wjhuang2016) [inline br/pkg/gluetidb/glue.go:128]: No close()?

File: br/pkg/restore/client.go [hunk 4]
```diff
@@ -461,13 +514,26 @@ func (rc *Client) CreatePolicies(ctx context.Context, policyMap *sync.Map) error
 	return err
 }
 
+// GetDBSchema gets the schema of a db from TiDB cluster
+func (rc *Client) GetDBSchema(dom *domain.Domain, dbName model.CIStr) (*model.DBInfo, error) {
+	info := dom.InfoSchema()
+	dbInfo, exist := info.SchemaByName(dbName)
+	if !exist {
+		log.Warn("schema not exist", zap.String("dbName", dbName.String()))
```
Comment (wjhuang2016) [inline br/pkg/restore/client.go:522]:
```suggestion
		log.Warn("schema does not exist", zap.String("dbName", dbName.String()))
```

File: br/pkg/restore/client.go [hunk 5]
```diff
@@ -461,13 +514,26 @@ func (rc *Client) CreatePolicies(ctx context.Context, policyMap *sync.Map) error
...
+}
+
 // CreateDatabase creates a database.
 func (rc *Client) CreateDatabase(ctx context.Context, db *model.DBInfo) error {
 	if rc.IsSkipCreateSQL() {
 		log.Info("skip create database", zap.Stringer("database", db.Name))
 		return nil
 	}
 
+	log.Info("create database", zap.Stringer("database", db.Name))
```
Comment (wjhuang2016) [inline br/pkg/restore/client.go:535]:
```suggestion
	log.Info("create database", zap.Stringer("name", db.Name))
```

File: br/pkg/restore/client.go [hunk 6]
```diff
@@ -1432,9 +1498,442 @@ func (rc *Client) PreCheckTableClusterIndex(
...
+			t := meta.NewMeta(txn)
+			ids, e = t.GenGlobalIDs(n)
+			return e
+		})
+
+	return ids, err
+}
+
+// UpdateSchemaVersion updates schema version by transaction way.
+func (rc *Client) UpdateSchemaVersion(ctx context.Context) error {
```
Comment (wjhuang2016) [inline br/pkg/restore/client.go:1907]: It's not compatible with the concurrent DDL, we need to take care of it.

File: br/pkg/restore/client.go [hunk 7]
```diff
@@ -1432,9 +1498,442 @@ func (rc *Client) PreCheckTableClusterIndex(
...
+				OldName:  name,
+				NewName:  newDBInfo.Name.String(),
+				TableMap: make(map[stream.OldID]*stream.TableReplace),
+			}
+			dbMap[t.DB.ID] = dbReplace
+		}
+
+		dbReplace.TableMap[t.Info.ID] = &stream.TableReplace{
+			TableID:      newTableInfo.ID,
+			OldName:      t.Info.Name.String(),
```
[the remark this answers is not in the document]
Comment (joccau) [reply br/pkg/restore/client.go]: The field 'OldName' has been removed from the struct 'stream.TableReplace'

File: br/pkg/restore/util.go [hunk 8]
```diff
@@ -28,39 +29,140 @@ import (
...
+type AppliedFile interface {
+	GetStartKey() []byte
+	GetEndKey() []byte
+}
+
+// getTableIDMap creates a map maping old tableID to new tableID.
+func getTableIDMap(newTable, oldTable *model.TableInfo) map[int64]int64 {
+	tableIDMap := make(map[int64]int64)
+
+	tableIDMap[oldTable.ID] = newTable.ID
```
[the remark this answers is not in the document]
Comment (joccau) [reply br/pkg/restore/util.go]: Have modified it.
Use a temporary map( name mapping id) to  improve performance.

**Review by YuJuncen — APPROVED** [review]
Given all commits LGTM, trivially(?) this PR LGTM.

**Review by 3pointer — APPROVED** [review]
LGTM

**Pull request merged**